filecache/__init__.py

```diff
@@ -269,65 +391,112 @@
+    def _unprotected_retrieve(self, filename, local_path):
+        if local_path.exists():
+            if self._filecache.is_shared or self._filecache._is_cached(local_path):
+                return local_path
             raise FileExistsError(f'Internal error - File already exists: {filename}')
 
         local_path.parent.mkdir(parents=True, exist_ok=True)
 
         if self._source_type == 'web':
-            url = f'{self._prefix}{filename}'
-            try:
-                response = requests.get(url)
-            except (requests.exceptions.ConnectionError,
-                    requests.exceptions.ConnectTimeout):
-                raise FileNotFoundError(f'Failed to download file from: {url}')
-            if response.status_code == 200:
-                with open(local_path, 'wb') as f:
-                    f.write(response.content)
-                self._filecache._record_cached_file(filename, local_path)
-                return local_path
-            else:
-                raise FileNotFoundError(f'Failed to download file from: {url}')
+            return self._retrieve_from_web(filename, local_path)
 
         if self._source_type == 'gs':
-            blob_name = f'{self._prefix}{filename}'
-            blob = self._gs_bucket.blob(blob_name)
-            try:
-                blob.download_to_filename(str(local_path))
-            except (google.api_core.exceptions.BadRequest,  # bad bucket name
-                    google.resumable_media.common.InvalidResponse,  # bad bucket name
-                    google.cloud.exceptions.NotFound):  # bad filename
-                # The google API library will still create the file before noticing
-                # that it can't be downloaded, so we have to remove it here
-                local_path.unlink()
-                raise FileNotFoundError(
-                    f'Failed to download file from: gs://{self._gs_bucket_name}/'
-                    f'{blob_name}')
-            self._filecache._record_cached_file(filename, local_path)
-            return local_path
+            return self._retrieve_from_gs(filename, local_path)
 
         if self._source_type == 's3':
-            s3_key = f'{self._prefix}{filename}'
-            try:
-                self._s3_client.download_file(self._s3_bucket_name,
-                                              s3_key,
-                                              str(local_path))
-            except botocore.exceptions.ClientError:
-                raise FileNotFoundError(
-                    f'Failed to download file from: s3://{self._s3_bucket_name}/'
-                    f'{s3_key}')
-            self._filecache._record_cached_file(filename, local_path)
-            return local_path
+            return self._retrieve_from_s3(filename, local_path)
+
+        raise AssertionError(
+            f'Internal error unknown source type {self._source_type}')  # pragma: no cover
+
+    def _retrieve_from_web(self, filename, local_path):
```

Comment:
<div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/SETI/rms-filecache/pull/2#issuecomment-2377799277>#3be435</a><span> - 09/26/2024, 07:56 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The current implementation downloads the entire file content into memory before writing it to disk, which can be inefficient for large files.
</div>
<div id="fix">
<b>Fix:</b> Modify the _retrieve_from_web method to use streaming download and write the content to disk in chunks.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -436,13 +436,15 @@ class FileCacheSource:
      def _retrieve_from_web(self, filename, local_path):
          url = f'{self._prefix}{filename}'
          try:
 -            response = requests.get(url)
 +            response = requests.get(url, stream=True)
              response.raise_for_status()
          except requests.exceptions.RequestException as e:
              raise FileNotFoundError(f'Failed to download file from: {url}') from e

          with open(local_path, 'wb') as f:
 -            f.write(response.content)
 +            for chunk in response.iter_content(chunk_size=8192):
 +                if chunk:
 +                    f.write(chunk)

          self._filecache._record_cached_file(filename, local_path)
          return local_path
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged